diff --git a/include/configs/ap121.h b/include/configs/ap121.h
index 1bd274d..55f1499 100644
--- a/include/configs/ap121.h
+++ b/include/configs/ap121.h
@@ -359,6 +359,11 @@
 #define AR7240_ART_PCICFG_OFFSET        3
 
 
+#define CONFIG_RESET_TO_RETRY		1
+#define CONFIG_BOOT_RETRY_TIME		30
+#define CONFIG_AUTOBOOT_KEYED
+#define CONFIG_AUTOBOOT_PROMPT		"autoboot in %d seconds (type 'ard' to enter u-boot console)...\n"
+#define CONFIG_AUTOBOOT_STOP_STR	"ard"
 
 #include <cmd_confdefs.h>
 
