--- a/drivers/net/wireless/ath/ath9k/init.c
+++ b/drivers/net/wireless/ath/ath9k/init.c
@@ -477,8 +477,7 @@ static int ath9k_init_soc_platform(struc
 			return ret;
 	}
 
-	if (pdata->tx_gain_buffalo)
-		ah->config.tx_gain_buffalo = true;
+	ah->config.tx_gain_buffalo = true;
 
 	return ret;
 }
@@ -511,7 +510,7 @@ static int ath9k_init_softc(u16 devid, s
 	sc->tx99_power = MAX_RATE_POWER + 1;
 	init_waitqueue_head(&sc->tx_wait);
 
-	if (!pdata || pdata->use_eeprom) {
+	if (!pdata) {
 		ah->ah_flags |= AH_USE_EEPROM;
 		sc->sc_ah->led_pin = -1;
 	} else {
--- a/include/linux/ath9k_platform.h
+++ b/include/linux/ath9k_platform.h
@@ -33,15 +33,12 @@ struct ath9k_platform_data {
 
 	bool endian_check;
 	bool is_clk_25mhz;
-	bool tx_gain_buffalo;
 	bool disable_2ghz;
 	bool disable_5ghz;
 
 	int (*get_mac_revision)(void);
 	int (*external_reset)(void);
 
-	bool use_eeprom;
-
 	int num_leds;
 	const struct gpio_led *leds;
 };
--- a/drivers/net/wireless/ath/ath9k/pci.c
+++ b/drivers/net/wireless/ath/ath9k/pci.c
@@ -693,7 +693,7 @@ static bool ath_pci_eeprom_read(struct a
 	struct ath_softc *sc = (struct ath_softc *) common->priv;
 	struct ath9k_platform_data *pdata = sc->dev->platform_data;
 
-	if (pdata && !pdata->use_eeprom) {
+	if (pdata) {
 		if (off >= (ARRAY_SIZE(pdata->eeprom_data))) {
 			ath_err(common,
 				"%s: eeprom read failed, offset %08x is out of range\n",
